clang/test/CIR/CodeGen/volatile.cpp

```diff
@@ -0,0 +1,131 @@
+// RUN: %clang_cc1 -std=c++20 -triple x86_64-unknown-linux-gnu -fclangir-enable -emit-cir %s -o %t.cir
+// RUN: FileCheck --input-file=%t.cir %s
+
+int test_load(volatile int *ptr) {
+  return *ptr;
+}
+
+//      CHECK: cir.func @_Z9test_loadPVi
+// CHECK-NEXT:   %0 = cir.alloca !cir.ptr<!s32i>, cir.ptr <!cir.ptr<!s32i>>
+// CHECK-NEXT:   %1 = cir.alloca !s32i, cir.ptr <!s32i>, ["__retval"]
+// CHECK-NEXT:   cir.store %arg0, %0 : !cir.ptr<!s32i>, cir.ptr <!cir.ptr<!s32i>>
+// CHECK-NEXT:   %2 = cir.load deref %0 : cir.ptr <!cir.ptr<!s32i>>, !cir.ptr<!s32i>
+// CHECK-NEXT:   %3 = cir.load volatile %2 : cir.ptr <!s32i>, !s32i
+// CHECK-NEXT:   cir.store %3, %1 : !s32i, cir.ptr <!s32i>
+// CHECK-NEXT:   %4 = cir.load %1 : cir.ptr <!s32i>, !s32i
+// CHECK-NEXT:   cir.return %4 : !s32i
+
+void test_store(volatile int *ptr) {
+  *ptr = 42;
+}
+
+//      CHECK: cir.func @_Z10test_storePVi
+// CHECK-NEXT:   %0 = cir.alloca !cir.ptr<!s32i>, cir.ptr <!cir.ptr<!s32i>>
+// CHECK-NEXT:   cir.store %arg0, %0 : !cir.ptr<!s32i>, cir.ptr <!cir.ptr<!s32i>>
+// CHECK-NEXT:   %1 = cir.const(#cir.int<42> : !s32i) : !s32i
+// CHECK-NEXT:   %2 = cir.load deref %0 : cir.ptr <!cir.ptr<!s32i>>, !cir.ptr<!s32i>
+// CHECK-NEXT:   cir.store volatile %1, %2 : !s32i, cir.ptr <!s32i>
+// CHECK-NEXT:   cir.return
+
+struct Foo {
+  int x;
+  volatile int y;
+  volatile int z: 4;
+};
+
+int test_load_field1(volatile Foo *ptr) {
+  return ptr->x;
+}
+
+//      CHECK: cir.func @_Z16test_load_field1PV3Foo
+// CHECK-NEXT:   %0 = cir.alloca !cir.ptr<!ty_22Foo22>, cir.ptr <!cir.ptr<!ty_22Foo22>>
+// CHECK-NEXT:   %1 = cir.alloca !s32i, cir.ptr <!s32i>
+// CHECK-NEXT:   cir.store %arg0, %0 : !cir.ptr<!ty_22Foo22>, cir.ptr <!cir.ptr<!ty_22Foo22>>
+// CHECK-NEXT:   %2 = cir.load %0 : cir.ptr <!cir.ptr<!ty_22Foo22>>, !cir.ptr<!ty_22Foo22>
+// CHECK-NEXT:   %3 = cir.get_member %2[0] {name = "x"} : !cir.ptr<!ty_22Foo22> -> !cir.ptr<!s32i>
+// CHECK-NEXT:   %4 = cir.load volatile %3 : cir.ptr <!s32i>, !s32i
```

Comment:
Thanks for the testcases. Since the overall features are already tested elsewhere, you can probably just check the relevant bits here. Example:

```
//      CHECK: cir.func @_Z16test_load_field1PV3Foo
//      CHECK: cir.load volatile
//      CHECK-NOT: cir.store volatile
```

If you prefere a more complete testcase, please then add regexes for the values, example:

```
//      CHECK: cir.func @_Z16test_load_field1PV3Foo
//      CHECK: %[[MemberAddr:.*]] = cir.get_member
//      CHECK: cir.load volatile %[[MemberAddr]]
```

Reply:
Updated.